librad/src/git/p2p/server.rs

```diff
@@ -211,24 +211,31 @@ impl UploadPack {
         match self {
             Self::AdvertiseRefs(mut child) => {
                 let mut stdout = child.stdout.take().unwrap().compat();
-                let mut buf = vec![];
 
                 send.write_all(UPLOAD_PACK_HEADER).await?;
-                futures::try_join!(
-                    futures::io::copy(&mut stdout, &mut send),
-                    recv.read_to_end(&mut buf),
-                    child.wait(),
-                )
-                .and_then(|(_, _, status)| {
-                    if !status.success() {
-                        Err(io::Error::new(
-                            io::ErrorKind::Other,
-                            format!("upload-pack ls exited non-zero: {:?}", status),
-                        ))
-                    } else {
-                        Ok(())
-                    }
-                })
+                let (_, status) =
+                    futures::try_join!(futures::io::copy(&mut stdout, &mut send), child.wait())?;
+
+                // Drive the `RecvStream` to its end to ensure all control messages are received
+                // and corresponding responses are scheduled. If this is not done crucial finish
+                // transitions are missing which determine if control messages to increase the
+                // mutual max stream counter are sent. It will manifest in one end not being
+                // able to open new streams, which in turn is not surfaced as
+                // hard error at this point by the quic lib.
+                //
+                // As there is no data expected on the `RecvStream` the expectation is that this
+                // completes with an ordinary EOF.
+                let mut buf = [0; 0];
+                recv.read(&mut buf).await?;
```

Comment:
This won't happen if the `try_join` fails right? Should we save the `?` on the `try_join` result until after running `recv.read`?